src/backend/distributed/executor/multi_executor.c

```diff
@@ -24,273 +24,440 @@
-			{
-				MultiTaskTrackerExecute(workerJob);
-			}
+	scanState->customScanState.methods = &TaskTrackerCustomExecMethods;
 
-			/* then create the result relation */
-			ProcessUtility((Node *) masterCreateStmt,
-						   "(temp table creation)",
-						   PROCESS_UTILITY_QUERY,
-						   NULL,
-						   None_Receiver,
-						   NULL);
+	return (Node *) scanState;
+}
 
-			/* make the temporary table visible */
-			CommandCounterIncrement();
 
-			if (!(eflags & EXEC_FLAG_EXPLAIN_ONLY))
-			{
-				CopyQueryResults(masterCopyStmtList);
-			}
+/*
+ * RouterCreateScan creates a custom scan node which sets callback functions
+ * for router executor depending on the router executor type.
+ */
+Node *
+RouterCreateScan(CustomScan *scan)
+{
+	CitusScanState *scanState = palloc0(sizeof(CitusScanState));
+	MultiPlan *multiPlan = NULL;
+	Job *workerJob = NULL;
+	List *taskList = NIL;
+
+	scanState->executorType = MULTI_EXECUTOR_ROUTER;
+	scanState->customScanState.ss.ps.type = T_CustomScanState;
+	scanState->multiPlan = GetMultiPlan(scan);
 
-			/*
-			 * Update the QueryDesc's snapshot so it sees the table. That's not
-			 * particularly pretty, but we don't have much of a choice.  One might
-			 * think we could unregister the snapshot, push a new active one,
-			 * update it, register it, and be happy. That only works if it's only
-			 * registered once though...
-			 */
-			queryDesc->snapshot->curcid = GetCurrentCommandId(false);
-
-			/*
-			 * Set the OID of the RTE used in the master select statement to point
-			 * to the now created (and filled) temporary table. The target
-			 * relation's oid is only known now.
-			 */
-			masterRangeTableEntry =
-				(RangeTblEntry *) linitial(masterSelectPlan->rtable);
-			masterRangeTableEntry->relid =
-				RelnameGetRelid(masterRangeTableEntry->eref->aliasname);
-
-			/*
-			 * Replace to-be-run query with the master select query. As the
-			 * planned statement is now replaced we can't call GetMultiPlan() in
-			 * the later hooks, so we set a flag marking this as a distributed
-			 * statement running on the master. That e.g. allows us to drop the
-			 * temp table later.
-			 *
-			 * We copy the original statement's queryId, to allow
-			 * pg_stat_statements and similar extension to associate the
-			 * statement with the toplevel statement.
-			 */
-			masterSelectPlan->queryId = queryDesc->plannedstmt->queryId;
-			queryDesc->plannedstmt = masterSelectPlan;
-
-			eflags |= EXEC_FLAG_CITUS_MASTER_SELECT;
+	multiPlan = scanState->multiPlan;
+	workerJob = multiPlan->workerJob;
+	taskList = workerJob->taskList;
+
+	/* check if this is a single shard query */
+	if (list_length(taskList) == 1)
+	{
+		bool isModificationQuery = IsModifyMultiPlan(multiPlan);
+		if (isModificationQuery)
+		{
+			scanState->customScanState.methods = &RouterSingleModifyCustomExecMethods;
+		}
+		else
+		{
+			scanState->customScanState.methods = &RouterSelectCustomExecMethods;
```

Comment:
Does it make sense to always check `isModificationQuery` so we can put an `Assert(isModificationQuery)` inside this block? Or is it possible to have `SELECTs` with multiple tasks that somehow reach the router executor here?

If so, I think the custom exec methods are misnamed here, since they imply they're for modification only. If not, I think we should add an `Assert`.